primitiv/c/graph.cc

```diff
@@ -7,224 +7,193 @@
-}
-
-primitiv_Shape *primitiv_Graph_get_shape(const primitiv_Graph *graph,
-                                         const primitiv_Node *node) {
-  return to_c_from_value(to_cc(graph)->get_shape(*to_cc(node)));
-}
-primitiv_Shape *safe_primitiv_Graph_get_shape(const primitiv_Graph *graph,
-                                              const primitiv_Node *node,
-                                              primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Graph_get_shape(graph, node), status, nullptr);
-}
-
-primitiv_Device *primitiv_Graph_get_device(const primitiv_Graph *graph,
-                                           const primitiv_Node *node) {
-  return to_c(&to_cc(graph)->get_device(*to_cc(node)));
-}
-primitiv_Device *safe_primitiv_Graph_get_device(const primitiv_Graph *graph,
-                                                const primitiv_Node *node,
-                                                primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Graph_get_device(graph, node), status, nullptr);
-}
-
-char *primitiv_Graph_dump(const primitiv_Graph *graph, const char *format) {
-  std::string str = to_cc(graph)->dump(format);
-  uint64_t len = str.length();
-  auto *c = new char[len + 1];
-  std::strncpy(c, str.c_str(), len);
-  return c;
-}
-char *safe_primitiv_Graph_dump(const primitiv_Graph *graph,
-                               const char *format,
-                               primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Graph_dump(graph, format), status, nullptr);
+  to_cpp_ptr(graph)->clear();
+}
+
+primitiv_Status primitiv_Graph_forward(primitiv_Graph *graph,
+                                       const primitiv_Node *node,
+                                       const primitiv_Tensor **tensor) {
+  try {
+    *tensor = to_c_ptr(&to_cpp_ptr(graph)->forward(*to_cpp_ptr(node)));
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
+}
+
+primitiv_Status primitiv_Graph_backward(primitiv_Graph *graph,
+                                        const primitiv_Node *node) {
+  try {
+    to_cpp_ptr(graph)->backward(*to_cpp_ptr(node));
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
+}
+
+primitiv_Status primitiv_Graph_get_shape(const primitiv_Graph *graph,
+                                         const primitiv_Node *node,
+                                         primitiv_Shape **shape) {
+  try {
+    *shape = to_c_ptr_from_value(
+        to_cpp_ptr(graph)->get_shape(*to_cpp_ptr(node)));
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
+}
+
+primitiv_Status primitiv_Graph_get_device(const primitiv_Graph *graph,
+                                          const primitiv_Node *node,
+                                          primitiv_Device **device) {
+  try {
+    *device = to_c_ptr(&to_cpp_ptr(graph)->get_device(*to_cpp_ptr(node)));
+    return ::primitiv_Status::PRIMITIV_OK;
+  } HANDLE_EXCEPTION
+}
+
+primitiv_Status primitiv_Graph_dump(
```

Comment:
This function has the same problem as `primitiv_Shape_to_string()`. See comments of that function.